datafusion/physical-plan/src/recursive_query.rs

```diff
@@ -0,0 +1,361 @@
+    name: String,
+    /// The dynamic part (recursive term) as is (without being executed)
+    recursive_term: Arc<dyn ExecutionPlan>,
+    /// The static part (static term) as a stream. If the processing of this
+    /// part is completed, then it will be None.
+    static_stream: Option<SendableRecordBatchStream>,
+    /// The dynamic part (recursive term) as a stream. If the processing of this
+    /// part has not started yet, or has been completed, then it will be None.
+    recursive_stream: Option<SendableRecordBatchStream>,
+    /// The schema of the output.
+    schema: SchemaRef,
+    /// In-memory buffer for storing a copy of the current results. Will be
+    /// cleared after each iteration.
+    buffer: Vec<RecordBatch>,
+    // /// Metrics.
+    _baseline_metrics: BaselineMetrics,
+}
+
+impl RecursiveQueryStream {
+    /// Create a new recursive query stream
+    fn new(
+        task_context: Arc<TaskContext>,
+        name: String,
+        recursive_term: Arc<dyn ExecutionPlan>,
+        static_stream: SendableRecordBatchStream,
+        baseline_metrics: BaselineMetrics,
+    ) -> Self {
+        let schema = static_stream.schema();
+        Self {
+            task_context,
+            name,
+            recursive_term,
+            static_stream: Some(static_stream),
+            recursive_stream: None,
+            schema,
+            buffer: vec![],
+            _baseline_metrics: baseline_metrics,
+        }
+    }
+
+    /// Push a clone of the given batch to the in memory buffer, and then return
+    /// a poll with it.
+    fn push_batch(
+        mut self: std::pin::Pin<&mut Self>,
+        batch: RecordBatch,
+    ) -> Poll<Option<Result<RecordBatch>>> {
+        self.buffer.push(batch.clone());
+        Poll::Ready(Some(Ok(batch)))
+    }
+
+    /// Start polling for the next iteration, will be called either after the static term
+    /// is completed or another term is completed. It will follow the algorithm above on
+    /// to check whether the recursion has ended.
+    fn poll_next_iteration(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<RecordBatch>>> {
+        let total_length = self
+            .buffer
+            .iter()
+            .fold(0, |acc, batch| acc + batch.num_rows());
+
+        if total_length == 0 {
+            return Poll::Ready(None);
+        }
+
+        // The initial capacity of the channels is the same as the number of partitions
+        // we currently hold in the buffer.
+        let (sender, receiver): (
+            Sender<Result<RecordBatch>>,
+            Receiver<Result<RecordBatch>>,
+        ) = mpsc::channel(self.buffer.len() + 1);
+
+        // There shouldn't be any handlers with this name, since the execution of recursive
+        // term will immediately consume the relation handler.
+        self.task_context
+            .push_relation_handler(self.name.clone(), receiver)?;
```

Comment:
The CTE may be referenced or selected multiple times. 
I tested the following queries on datafusion-cli, and they failed.
```sh
DataFusion CLI v33.0.0
❯ CREATE TABLE t(a BIGINT) AS VALUES(1),(2),(3);
0 rows in set. Query took 0.007 seconds.

❯ WITH RECURSIVE my_cte AS
      (SELECT a from t UNION ALL SELECT a+2 FROM my_cte WHERE a<5)
   SELECT * FROM my_cte t1, my_cte;
External error: Internal error: Relation handler my_cte already registered.
This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker
```

Reply:
Thinking through the best way to fix this...

Reply:
I added logic to wrap the logical plan with aliases provided by sqlparser to mirror what's done along the regular CTE path. The resulting plan now looks like 

```text
logical_plan
CrossJoin:
--SubqueryAlias: t1
----SubqueryAlias: my_cte
------RecursiveQuery: is_distinct=false
--------TableScan: t projection=[a]
--------Projection: t.a + Int64(2) AS a
----------Filter: t.a < Int64(5)
------------NamedRelation: my_cte
--SubqueryAlias: my_cte
----RecursiveQuery: is_distinct=false
------TableScan: t projection=[a]
------Projection: t.a + Int64(2) AS a
--------Filter: t.a < Int64(5)
----------NamedRelation: my_cte
physical_plan
CrossJoinExec
--RecursiveQueryExec: is_distinct=false
----MemoryExec: partitions=1, partition_sizes=[1]
----CoalescePartitionsExec
------ProjectionExec: expr=[a@0 + 2 as a]
--------CoalesceBatchesExec: target_batch_size=8192
----------FilterExec: a@0 < 5
------------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
--------------ContinuanceExec: name=my_cte
--RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
----RecursiveQueryExec: is_distinct=false
------MemoryExec: partitions=1, partition_sizes=[1]
------CoalescePartitionsExec
--------ProjectionExec: expr=[a@0 + 2 as a]
----------CoalesceBatchesExec: target_batch_size=8192
------------FilterExec: a@0 < 5
--------------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
----------------ContinuanceExec: name=my_cte
```

But still getting the same error because we need a way to inform each branch of the cross join to incorporate the alias in the recursive query and named relation physical plan translations. I'm thinking that I could hijack the planning context to accomplish this, perhaps.

Reply:
I'd also be open to pushing the error into the logical plan build time (preventing this behavior for now until we implement support). But both approaches are probably the same amount of work, so we should just get it working.

Reply:
Fixed the issue @jonahgao in this [commit](https://github.com/apache/arrow-datafusion/pull/7581/commits/d7721f12d317f31c94712e89813abbf7d7ec44d7). But the fix might be a little invasive so perhaps there is a simpler way to pass down context at logical or physical plan build time.